### PR TITLE
[BEAM-12000] Add Python 3.9 support.

### PR DESCRIPTION
**Pull request by tvalentyn**
Add Python 3.9 infrastructure and test suites for Apache Beam.

### CODE REVIEW

**Comment by tvalentyn** [comment]
run seed job

**Comment by tvalentyn** [comment]
Run Seed Job

**Comment by tvalentyn** [comment]
Run Seed Job

**Comment by tvalentyn** [comment]
Run Python 3.9 PostCommit

**Comment by tvalentyn** [comment]
Run Dataflow Python ValidatesRunner

**Comment by tvalentyn** [comment]
Run Python Dataflow ValidatesRunner

**Comment by aaltay** [comment]
What is the next step on this PR?

**Comment by tvalentyn** [comment]
Next step  is to  remove changes related to Dataflow Py 3.9 support since that runner needs additional work, and check in bits related to Apache Beam 3.9 support.

**Comment by robertwb** [comment]
Is this still being worked on? Are there parts of this CL that we can check in sooner rather than later?

**Comment by tvalentyn** [comment]
yes, it's being worked on as we speak.

**Comment by tvalentyn** [comment]
Thanks for taking a look at BEAM-12914. I will exclude the 4 failing test temporarily in this PR and it should be ready to go.

**Comment by tvalentyn** [comment]
R: @AnandInguva

**Comment by tvalentyn** [comment]
Cython currently uninstallable on Py3.9 on Jenkins VMs after Ubuntu upgrade, looking into that.

File: build.gradle.kts [hunk 1]
```diff
@@ -278,9 +279,11 @@ tasks.register("pythonDockerBuildPreCommit") {
   dependsOn(":sdks:python:container:py36:docker")
   dependsOn(":sdks:python:container:py37:docker")
   dependsOn(":sdks:python:container:py38:docker")
+  dependsOn(":sdks:python:container:py39:docker")
 }
 
 tasks.register("pythonLintPreCommit") {
+  // TODO(BEAM-8890): Find a better way to specify lint and formatter tasks without hardcoding py version.
```
Comment (AnandInguva) [inline build.gradle.kts:286]: Just to confirm, for this Jira, I see an issue related to  updating `google-cloud-java`. I don't see anything on lint and formatter.

File: build.gradle.kts [hunk 2]
```diff
@@ -313,15 +316,23 @@ tasks.register("python38PostCommit") {
...
+  dependsOn(":sdks:python:test-suites:dataflow:py39:postCommitIT")
+  dependsOn(":sdks:python:test-suites:direct:py39:postCommitIT")
+  dependsOn(":sdks:python:test-suites:direct:py39:hdfsIntegrationTest")
+  dependsOn(":sdks:python:test-suites:portable:py39:postCommitPy39")
+}
+
 tasks.register("portablePythonPreCommit") {
   dependsOn(":sdks:python:test-suites:portable:py36:preCommitPy36")
-  dependsOn(":sdks:python:test-suites:portable:py37:preCommitPy37")
+  dependsOn(":sdks:python:test-suites:portable:py39:preCommitPy39")
```
Comment (AnandInguva) [inline build.gradle.kts:328]: why do we have only py36, py39?

File: build.gradle.kts (hunk 1 above)
Comment (AnandInguva) [reply build.gradle.kts:286]: I think it should be BEAM-9980

File: build.gradle.kts (hunk 2 above)
Comment (tvalentyn) [reply build.gradle.kts:328]: For some test suites, the main value may be in testing the highest supported version and the lowest supported version only to save resources and reduce execution time. But, existing suite definitions not always follow this rule and sometimes test random versions. Some suites test all versions.

File: build.gradle.kts (hunk 1 above)
Comment (tvalentyn) [reply build.gradle.kts:286]: good catch

**Comment by tvalentyn** [comment]
Fixed one more issue (https://issues.apache.org/jira/browse/BEAM-13845 has details)

File: .github/workflows/python_tests.yml [hunk 3]
```diff
@@ -173,12 +174,12 @@ jobs:
...
       matrix:
         os: [ubuntu-latest, macos-latest, windows-latest]
-        python: [3.6, 3.7, 3.8]
+        python: [3.6, 3.7, 3.8, 3.9]
         exclude:
           # TODO remove exclusion after issue with protobuf is solved
           # https://github.com/protocolbuffers/protobuf/issues/7765
           - os: windows-latest
-            python: 3.8
+            python: 3.9
```
Comment (AnandInguva) [inline .github/workflows/python_tests.yml:182]: According the PR, this is solved for 3.9[Wheels are now present for Windows] IIUC. Can you check the PR and if it solved, can we remove the exclude for Python 3.9?

File: build.gradle.kts [hunk 4]
```diff
@@ -347,12 +358,14 @@ tasks.register("runBeamDependencyCheck") {
 }
 
 tasks.register("whitespacePreCommit") {
+  // TODO(BEAM-8890): Find a better way to specify the tasks without hardcoding py version.
```
[the remark this answers is not in the document]
Comment (tvalentyn) [reply build.gradle.kts]: Done

File: .github/workflows/python_tests.yml [hunk 5]
```diff
@@ -140,12 +141,12 @@ jobs:
...
```
[the remark this answers is not in the document]
Comment (tvalentyn) [reply .github/workflows/python_tests.yml]: Thanks for catching!

**Comment by tvalentyn** [comment]
Thanks, PTAL, @AnandInguva

**Review by AnandInguva — APPROVED** [review]

**Comment by AnandInguva** [comment]
LGTM!

**Comment by tvalentyn** [comment]
Run Seed Job

**Comment by tvalentyn** [comment]
Run Python 3.9 PostCommit

**Comment by tvalentyn** [comment]
Run Python Dataflow V2 ValidatesRunner

**Comment by tvalentyn** [comment]
Run Python Dataflow ValidatesRunner

**Comment by tvalentyn** [comment]
Run Python Dataflow ValidatesContainer

**Comment by tvalentyn** [comment]
Run Python Examples_Direct

**Comment by tvalentyn** [comment]
Run Python 3.9 PostCommit

**Comment by tvalentyn** [comment]
Run Python 3.9 PostCommit

**Comment by tvalentyn** [comment]
Run Seed Job

**Comment by tvalentyn** [comment]
Merging based on the result of local execution + prior test results.

**Pull request merged**